src/poll/responser.rs

```diff
@@ -0,0 +1,666 @@
+                        }
+                    };
+
+                    let sge = Sge {
+                        addr: resp.desc.laddr,
+                        len: resp.desc.len,
+                        key: resp.desc.lkey,
+                    };
+                    let desc_builder = ToCardWorkRbDescBuilder::new_read_resp()
+                        .with_common(common)
+                        .with_sge(sge);
+                    if let Err(e) = device.push(desc_builder) {
+                        eprintln!("Failed to push read response: {:?}", e);
+                    }
+                }
+                Err(_) => {
+                    // The only error is pipe broken, so just exit the thread
+                    return;
+                }
+            }
+        }
+    }
+}
+
+type Slot = [u8];
+
+/// A structure to hold the acknowledge buffer
+///
+/// TODO: currently, it does not support the auto buffer recycling.
+pub(crate) struct AcknowledgeBuffer {
+    free_list: Queue<Option<&'static mut Slot>>,
+    start_va: usize,
+    length: usize,
+    lkey: u32,
+}
+
+impl AcknowledgeBuffer {
+    #[inline]
+    pub fn get_lkey(&self) -> u32 {
+        self.lkey
+    }
+
+    /// Create a new acknowledge buffer
+    pub fn new(start_va: usize, length: usize, lkey: u32) -> Self {
+        assert!(length % ACKNOWLEDGE_BUFFER_SLOT == 0);
+        let free_list = Queue::new();
+        let mut va = start_va;
+        let slots: usize = length / ACKNOWLEDGE_BUFFER_SLOT;
+
+        for _ in 0..slots {
+            // SAFETY: the buffer given by the user should be valid, can be safely converted to array
+            let buf = unsafe { from_raw_parts_mut(va as *mut u8, ACKNOWLEDGE_BUFFER_SLOT) };
+            free_list.push(Some(buf));
+            va += ACKNOWLEDGE_BUFFER_SLOT;
+        }
+        Self {
+            free_list,
+            start_va,
+            length,
+            lkey,
+        }
+    }
+
+    pub fn alloc(&self) -> Option<&'static mut Slot> {
+        // FIXME: currently, we just recycle all the buffer in the free list.
+        let result = self.free_list.pop();
+        match result {
+            Some(Some(buf)) => Some(buf),
+            Some(None) => None,
+            None => {
+                // The buffer is already freed, so we just try to allocate another buffer
+                let mut va = self.start_va;
+                let slots: usize = self.length / ACKNOWLEDGE_BUFFER_SLOT;
+                for _ in 0..slots {
+                    // SAFETY: the buffer given by the user should be valid, can be safely converted to array
+                    let buf = unsafe { from_raw_parts_mut(va as *mut u8, ACKNOWLEDGE_BUFFER_SLOT) };
+                    self.free_list.push(Some(buf));
+                    va += ACKNOWLEDGE_BUFFER_SLOT;
+                }
+                self.free_list.pop().map(|x| x.unwrap())
+            }
+        }
+    }
+
+    pub fn free(&self, buf: &'static mut Slot) {
```

Comment:
shoud we use drop instead of free?

Reply:
If we implement a `drop` for the `Slot`, every slot will be freed after we send it to the `descriptor builder`. However, there is still a chance that the hardware does not consume the buffer while it has been used by other packets.

The design of the queue is a FIFO structure. The used buffer will be put at the end of the queue. Since we know the speed of hardware consumption is much faster than the feeding speed of the software, and we have a large enough queue so that we can guarantee that the recently used buffer will not be used again.  It makes sense.

I will update the design above with `Drop` and let it free after the thread sends the `descriptor`

Reply:
if `free` not take ownership, then it can be freed many times?

Reply:
Yes, you are right.  But it seems that I can not free the slot with only `Drop` in the current situation.
Recycling the slot requires a reference to the allocator's free list.  Maybe I should add an `Arc` pointer inside the `Slot` struct so that when the slot is dropped, it automatically adds itself to the free list?